[이노] 미션 4: 모든 기물 배치하기

eNoLJ opened the pull request:
이번 미션은 중복을 제거하는 데 중점을 뒀습니다.

1. Piece class에 말을 생성하는 메소드를 Color와 Representation enum을 인자로 받는 createPiece 메소드로 통일 했습니다.

2. Board class에 pieces 변수의 구조를 바꿨습니다. 기존 Map의 경우 color를 키로 pawn을 값으로 받았지만 pawn 이외의 다른 말들도 추가하기 위해 color를 키로 갖는 Map 안에 pawn과 piece를 키로 갖는 Map으로 변경 했습니다. 이 부분때문에 생성자 메소드와 그 외의 다른 메소드들이 좀 복잡해졌습니다. 현재까지는 괜찮은것 같으나 앞으로 어떻게 될지 약간 걱정이 되네요.

3. PieceTest class에는 기본생성자 테스트는 삭제하고 create 메소드에 폰 이외의 다른 말들을 테스트하는 코드를 추가 했습니다. 이 부분에서 6개의 말, 2개의 색으로 12줄이 중복되어 Representation enum의 값들로 테스트를 진행하도록 했습니다.

4. BoardTest class에는 기존에 불필요해진 테스트들은 제거하고 보드가 초기화 시 말의 개수와 배치를 테스트하는 코드를 작성했습니다.

냉정한 조언 부탁드립니다. 최대한 한다고 했는데 이번 미션에서는 코드들이 좀 복잡해진 감이 있어서 그 부분이 좀 걸리네요.
이번 리뷰도 잘 부탁드립니다.
항상 잘 배우고 있습니다!!
감사합니다.

eNoLJ committed:
Divide Piece enum into Color and Representation enum
Piece enum이 color와 representation을 담기에는 직관적이지 않아 Piece enum을 Color와 Representation enum으로 나눴습니다.

eNoLJ committed:
Refactor Pawn, PawnTest, Board, BoardTest class
Piece enum이 Color와 Representation enum으로 나뉨에 따라 코드 리팩토링

eNoLJ committed:
Create StringUtils class
운영체제별로 줄바꿈 문자가 다르기 때문에 System.getProperty("line.separator")를 StringUtils class에 상수로 지정, appendNewLine 메소드를 통해 문자열에 줄바꿈 문자를 추가하는 기능을 제공, 해당 기능을 제공하기 위해 StringUtils class 생성

eNoLJ committed:
Refactor print method in Board class
Board class 안에 있는 print 메소드에 줄바꿈 문자를 운영체제별로 동일하게 작동시키기 위해 앞서 만든 StringUtils class의 appendNewLine 메소드를 활용하여 리팩토링

eNoLJ committed:
Rename Pawn class to Piece class
chess에 폰뿐만 아니라 다른 말들을 만들기 위해 Pawn class를 Piece class로 이름 변경

eNoLJ committed:
Add KNIGHT, ROOK, BISHOP, QUEEN, KING constant in Representation enum
폰 이외의 다른 말들의 상수를 사용하기 위해 Representation enum에 나이트, 룩, 비숍, 퀸, 킹의 상수 추가

eNoLJ committed:
Refactor part about Representation in Board, Piece, BoardTest, PieceTest class
Representation enum의 변경에 따라 Board, Piece, BoardTest, PieceTest class 리팩토링

eNoLJ committed:
Create createPiece method in Piece class
Piece class를 값 객체로 취급하기 위해 생성자 함수를 private로 만들고 별도의 메소드로 객체를 생성하기 위해 createPiece 메소드 생성

eNoLJ committed:
Fix Board and BoardTest class
앞서 Piece class의 생성자를 private로 만들었기 때문에 생기는 오류를 createPiece 메소드를 통해 해결

eNoLJ committed:
Add another pieces test in PieceTest class
폰 이외의 다른 말들이 잘 생성되는지 확인을 위한 테스트 케이스 추가

eNoLJ committed:
Add piece by piece test in PieceTest class
폰 이외의 다른 말들까지 테스트하기 위해 PieceTest class의 create 메소드에 테스트 추가

eNoLJ committed:
Create a test to test the overall health of the board
폰 이외의 다른 말들을 보드에 초기화 하기 위해 기존에 불필요한 테스트들은 삭제하고 초기화 된 보드의 개수와 배치와 출력을 테스트하는 메소드 생성

eNoLJ committed:
Fix the Board class to initialize other pieces other than the pawn
폰 이외의 다른 말들을 초기화하기 위해 Board class의 pieces의 구조와 관련 메소드들을 수정

eNoLJ committed:
Refactor the Game class according to the method change of the Board class
Board class의 print 메소드가 showBoard 메소드로 변경 됨에 따라 Game class 리팩토링

eNoLJ committed:
Create color check method in Piece class
Piece 인스턴스의 색을 확인하기 위해 isWhite와 isBlack 메소드 생성

eNoLJ committed:
Refactor initialize method in Board class
Board class의 initialize 메소드의 중복되는 부분을 제거하기 위해 리팩토링

eNoLJ committed:
Change the order of method in Board class

wheejuni commented on src/main/java/net/eno/pieces/Representation.java:
```diff
@@ -0,0 +1,24 @@
+package net.eno.pieces;
+
+public enum Representation {
```
`Representation` 이라는 이름보다는 더 구체적이고 직접적인 이름이면 좋겠습니다.
기물의 종류와 종류에 따른 기호를 담고 있는 `enum` 인데요, "종류" 에 집중하면 더 좋은 선택이 될 것 같습니다.

eNoLJ replied:
`Representation`을 `PieceType`으로 이름을 변경했습니다.

wheejuni commented on src/main/java/net/eno/pieces/Piece.java:
```diff
@@ -1,24 +1,33 @@
 
     private final String color;
-    private final char pawn;
+    private final char representation;
```
`color` 와 `representation` 을 따로 담지 말고, 밑에 작성된 (이름이 변경되어야 할) `Representation` 이라는 `enum` 하나로도 해결할 수 있을 것 같습니다.

eNoLJ replied:
`Color`와 `Representation`를 하나의 enum으로 관리를 해야할지 나눠야할지 미션3에서도 고민을 많이 했었지만 피드백 결과에 따라 나누게 되었습니다. 제가 같이 사용할 수 있는 enum에 대해 마땅한 생각이 떠오르지 않더라구요. 아직도 고민 중인 부분입니다ㅠㅠ

아직 적용은 하지 않았지만 `Color`와 `Representation`을 `PieceType enum`으로 만들어 봤습니다. 조언해 주신 의도가 이게 맞는건지 잘 모르겠습니다. 오전 내내 다른 방법도 생각을 해보고 시도해봤지만 밑에 방법이 그나마 제일 나은 방법 같습니다. 하지만 밑에 방법으로써는 위에처럼 `Color`와 `Representation`을 나눴을 때보다 오히려 더 직관적이지 않은 것 같고 `Piece`를 생성할 때 `PieceType` 하나만을 받는걸로 구현한다면 `Piece`의 멤버변수로 `WHITE('p', 'r', 'n', 'b', 'q', 'k')`나  `BLACK('P', 'R', 'N', 'B', 'Q', 'K')`가 들어갈텐데 그렇게 된다면 모든 기물의 `representation`이 포함되어 있어서 `Piece`의 의미가 흔들리는 느낌이 듭니다. 그렇다고 `PieceType`과 `String piece`로 해당 기물을 문자로 함께 받자니 너무 별로에요....... 이부분 좀 더 조언을 해주시면 고쳐보도록 하겠습니다. 그 전까진 일단 밑에 코멘트들을 수정하고 있겠습니다.

```java
public enum PieceType {

    WHITE('p', 'r', 'n', 'b', 'q', 'k'),
    BLACK('P', 'R', 'N', 'B', 'Q', 'K');

    private final char pawn;
    private final char rook;
    private final char knight;
    private final char bishop;
    private final char queen;
    private final char king;

    PieceType(char pawn, char rook, char knight, char bishop, char queen, char king) {
        this.pawn = pawn;
        this.rook = rook;
        this.knight = knight;
        this.bishop = bishop;
        this.queen = queen;
        this.king = king;
    }

    public char getRepresentation(String piece) {
        switch (piece) {
            case "pawn" :
                return this.pawn;
            case "rook" :
                return this.rook;
            case "knight" :
                return this.knight;
            case "bishop" :
                return this.bishop;
            case "queen" :
                return this. queen;
            case "king" :
                return this.king;
            default :
                throw new InvalidParameterException("해당 기물은 존재하지 않습니다.");
        }
    }

}
```

tmdgusya replied:
음 위와 같이 작성하는건 조금 비효율적으로 생각됩니다..
getRepresentation 에서 switch 보다는 차라리 List 로 관리해서 stream filter 를 이용하는게 나은 방식으로 보입니다.

eNoLJ replied:
@tmdgusya 그런식으로도 해볼 수 있겠군요. 의견 감사합니다. 만약 위에 코드처럼 Color와 Representation을 하나의 enum으로 관리하게된다면 로치의 의견을 반영하도록 하겠습니다!!

아 그리고 혹시 로치는 Color와 Representation enum을 따로 만들거나 같이 만드는 부분에 있어서 어떻게 생각하시는지 말씀해주실 수 있으면 부탁드리겠습니다!!

wheejuni replied:
@eNoLJ 너무 어렵게 생각하시는 것 같은데요....
`WHITE_PAWN`, `BLACK_QUEEN` 과 같이 색깔과 타입을 모든 경우의 수로 풀어서 enum을 만들면 되지 않을까요.

eNoLJ replied:
아하 그런 의미였군요. 사고방식이 너무 유연하지 못했네요. 그런식으로 관리를 하면 확연히 좋아질 듯 합니다. 조언 감사드립니다!! 5단계에서는 그렇게 적용해 보겠습니다.

wheejuni commented on src/main/java/net/eno/pieces/Piece.java:
```diff
@@ -1,24 +1,33 @@
-    private final char pawn;
+    private final char representation;
+
+    private Piece(Color color, Representation representation) {
+        this.color = color.toString();
+        this.representation = representation.get(color);
+    }
```
생성자 파라메터를 잘 받아놓고, 굳이 각 파라메터에서 정보를 직접 꺼내서 객체의 필드에 다시 담는 구현은 조금 아쉬운 것 같습니다. 
파라메터를 그대로 살리면서 객체에 필요한 정보를 보관할 수 있는 방법을 다시 고민해주셔야 할 것 같네요.
이를 위해서는 `Color` 와 `Representation` 부터 다시 설계되어야 하겠죠. 더 고민해주세요.

eNoLJ replied:
`Piece`의 멤버변수 타입을 `Color`와 `Representation`으로 변경했습니다. 생성자에서 번거롭게 toString과 get으로 정보를 꺼내서 할당하는 문제를 해결했습니다.

wheejuni commented on src/main/java/net/eno/pieces/Piece.java:
```diff
@@ -1,24 +1,33 @@
+        return this.representation;
+    }
+
+    public boolean isWhite() {
+        return this.color.equals(Color.WHITE.toString());
+    }
+
+    public boolean isBlack() {
+        return this.color.equals(Color.BLACK.toString());
```
필드 타입이 `Color` 였다면 굳이 `.equals()` 나 `.toString()` 을 호출할 필요도 없었을 겁니다. 아쉽네요.

eNoLJ replied:
위 코멘트에 변경된 사항으로 인해 toString을 제거했습니다. 처음 멤버변수를 만들때 딱 String color 하나와 char representation 하나만을 가지게 하려는 의도로 저렇게 코드를 짰던 것인데 오히려 생성자와 isWhite, isBlack 메서드에서 정보를 꺼내오는 번거로움이 생기게 되었네요. 이부분은 전혀 의식하지 못했는데 피드백 감사드립니다.

wheejuni commented on src/main/java/net/eno/pieces/Representation.java:
```diff
@@ -0,0 +1,24 @@
+        this.black = black;
+    }
+
+    public char get(Color color) {
+        return color.toString().equals("WHITE") ? this.white : this.black;
+    }
```
색깔에 따른 디스플레이 캐릭터를 리턴하는 메소드인데 `.get()` 이라는 메소드명은 너무 모호합니다.

eNoLJ replied:
처음 의도했을 때 `Representation enum`을 사용하면 represeantation의 정보를 갖고 있다는걸 알 것 같아서 `get()`만으로 충분히 의미가 전달될 것이라 생각했습니다. 하지만 위에서 `PieceType`으로 enum의 이름이 변경되면서 `get()` 메소드를 `getRepresentation()`으로 이름을 변경해주었습니다.

wheejuni commented on src/main/java/net/eno/chess/Board.java:
```diff
@@ -1,68 +1,100 @@
     }
 
-    public String getPawnsResult(Piece piece) {
+    private List<Representation> sortRepresentation() {
```
메소드명이 `.sortRepresentation()` 입니다만 정렬하는 로직은 아닌 것 같네요.

eNoLJ replied:
가만히 생각해보니 정렬보다는 배치라는 의미가 맞는 것 같습니다. `PieceType`들을 보드에 올리기 위해 순서대로 '배치'를 한다는 의미로 `arrangePiece()`라는 이름으로 변경했습니다.

wheejuni commented on src/main/java/net/eno/chess/Board.java:
```diff
@@ -1,68 +1,100 @@
 public class Board {
 
-    private final Map<String, List<Pawn>> pawns;
+    private final Map<String, Map<String, List<Piece>>> pieces;
```
`Piece` 가 자신의 색깔과 타입을 관리하는데 이렇게까지 구분해서 각 `Piece` 객체를 보관할 이유는 없을 것 같습니다.
이 필드를 `List<Piece>` 로 선언하고, 색깔과 타입에 따른 분기 처리가 필요하다면 `Piece` 에게 직접 물어보면 될 것 같습니다. 
이런 방향으로 리팩토링이 이뤄진다면 이 필드에 의존하고 있는 `.addPiece()` 나 `.pieceCount()`, `.initialize()` 모두 지금보다 훨씬 깔끔해지게 될 겁니다.

eNoLJ replied:
`Map<String, Map<String, List<Piece>>> pieces`를 `List<List<piece>>`로 변경했습니다. `List<Piece>`로 선언하라고 말씀하셨는데 그렇게 되면 현재 미션에서 32개의 기물을 순서대로 추가해줘야한다고 생각했습니다. 그럼 관리가 너무 힘들어질 것 같아 리스트 안에 리스트를 만들게 되었습니다. 처음 리스트는 체스의 rank를 표현하려 했습니다. 하지만 index값을 그대로 rank로 사용하여 rank가 0이 있다는점, 현재 상태로는 rank가 4줄 밖에 없다는 문제점을 숙지하고 있으며 미션5에서 8개의 rank를 class로 구현하려 생각중입니다.

wheejuni commented on src/main/java/net/eno/chess/Board.java:
```diff
@@ -1,68 +1,100 @@
+        this.pieces = new HashMap<>();
+        this.pieces.put(Color.WHITE.toString(), new HashMap<String, List<Piece>>() {
+            {
+                put("pawn", new ArrayList<>());
+                put("piece", new ArrayList<>());
```
`pawn` 과 `piece` 가 구분돼야 한다면 그 정보는 `Piece` 에 보관돼야 할 겁니다. 이렇게 맵으로 관리하지 않고요.

eNoLJ replied:
저도 이부분을 구현할 때 `pawn`과 다른 `piece`를 구분하는거에 있어서 꺼림찍한 느낌을 받았습니다. 미션 3에서 만들었던 map을 이용한 `pieces`구조에 생각이 사로잡혀 새로 뜯어고칠 생각을 하지 못한 것 같습니다. 현제는 List안에 구분 없이 추가했습니다.

wheejuni commented on src/main/java/net/eno/chess/Board.java:
```diff
@@ -1,68 +1,100 @@
-        for (int i = 0; i < pawnSize(piece); i++) {
-            result.append(findPawn(piece, i).getRepresentation());
-        }
+        result.append(appendNewLine(getPiecesResult(Color.BLACK, "piece")));
+        result.append(appendNewLine(getPiecesResult(Color.BLACK, "pawn")));
+        result.append(appendNewLine("........"));
+        result.append(appendNewLine("........"));
+        result.append(appendNewLine("........"));
+        result.append(appendNewLine("........"));
+        result.append(appendNewLine(getPiecesResult(Color.WHITE, "pawn")));
+        result.append(appendNewLine(getPiecesResult(Color.WHITE, "piece")));
```
만약 게임판이 뒤집히게 되면, 그러니까 백팀이 위로, 흑팀이 아래로 오는 상황이라면 대처할 수 있을까요?

eNoLJ replied:
그 부분은 미션 요구사항에 나와있지않아 생각을 하지 못했습니다. 해주신 피드백에 맞춰 `showBoard()`메소드를 `showBoard(Color color)`로 변경하여 파라미터로 색을 받게 만들었습니다. 색을 지정하면 그에 맞는 색이 전면에 배치되도록 수정하였습니다.

wheejuni reviewed (changes requested):
냉정한(?) 리뷰를 원하셔서 최대한 냉정하게 작성해 보았습니다. ㅎㅎ 💯 👍 
4단계 작성하시느라 수고 많으셨고요, 피드백이 좀 많은데요, 하나하나 살펴보시면서 궁금해지는 부분은 다시 물어보시고, 수정 진행해주세요.

eNoLJ commented:
이번 미션4를 진행하면서 불필요하거나 모호한것이 있다는 느낌은 받았지만 그게 정확히 무엇이고 어떻게 수정해야할지 갈피를 잡지 못했는데 피드백을 보니 기분이 좋네요!! 냉정한 피드백 정말 정말 감사드립니다!! 지금부터 하나씩 수정하고 코멘트 남겨가도록 하겠습니다ㅎㅎ

eNoLJ committed:
Change the type of the member variable in Piece class
Piece 생성자에서 Color와 Representation을 파라미터로 받아 정보를 꺼내서 멤버변수에 할당해준 기존 방식이 번거롭고 isWhite와 isBlack 메소드에서 또한 Color에 toString을 해줘야하는 문제가 생겨  멤버변수를 Color와 Representation 타입으로 변경

eNoLJ committed:
Change enum name from Representation to PieceType
기존에 Representation이라는 이름이 기물의 종류와 종류의 캐릭터를 담기에는 부적절하여 PieceType이라는 이름으로 변경

eNoLJ committed:
Refactor the code of the part that used Enum with the changed name
Representation enum의 이름이 PieceType으로 바뀜에 따라 Piece와 PieceTest class 리팩토링

eNoLJ committed:
Change method name from sortRepresentation to arrangePiece
Board class의 sortRepresentation 메소드의 이름이 메소드가 행하는 일에 어울리지 않아 arrangePiece로 이름 변경

eNoLJ committed:
Change data structure of member variables in Board class
Board class의 Pieces 멤버변수의 이름을 board로 변경, map을 사용한 구조가 너무 복잡하여 list로 리팩토링, 그에 따라 메소드들의 구조 변경, showBoard 메소드에 color를 인자로 받아 전면의 말이 흰색과 검은색을 선택해서 볼 수있도록 기능 추가

eNoLJ committed:
Add test case in BoardTest class
BoardTest class에 전면이 화이트로 선택된 보드와 검은색으로 선택된 보드의 테스트 추가

eNoLJ committed:
Refactor code in Game class
Board class에 showBoard 메소드의 변경에 따라 리팩토링

wheejuni commented:
@eNoLJ 아직 수정 진행중이신거죠?

eNoLJ commented:
@wheejuni 리뷰 해주신부분 반영해서 진행한 상황입니다 해주신 리뷰마다 코멘트 달고 궁금한 점 적었습니다 보시고 의견을 한번 더 주시면 다시 반영하겠습니다!!

tmdgusya commented on src/main/java/net/eno/chess/Board.java:
```diff
@@ -14,33 +15,33 @@
-        this.pawns.put(Piece.WHITE.getColor(), new ArrayList<>());
-        this.pawns.put(Piece.BLACK.getColor(), new ArrayList<>());
+        this.pawns.put(Color.WHITE.toString(), new ArrayList<>());
+        this.pawns.put(Color.BLACK.toString(), new ArrayList<>());
```
이부분은 toString() 을 재정의 하신것 같은데 name() 이라는 메소드로는 불가능했을까요?
제가 해당 부분 코드를 보지 못해서 한번 해보시고 답변 달아주시면 감사하겠습니다.

eNoLJ replied:
toString()을 따로 정의한건 아니고 있는거 사용했습니다!! 그리고 name()이라는 메소드는 있는 줄 몰랐네요 덕분에 알게 되었습니다. 그리고 해당 코드는 최신 상태에선 수정 된 상태라 최신 코드를 봐주시면 감사하겠습니다!!

tmdgusya commented on src/test/java/net/eno/chess/BoardTest.java:
```diff
@@ -14,37 +12,32 @@
+        assertThat(board.showBoard()).isEqualTo(
+                appendNewLine("RNBQKBNR") +
+                appendNewLine("PPPPPPPP") +
+                blankRank + blankRank + blankRank + blankRank +
```
이부분 개인적인 취향인데 blank 도 동일하게 한줄한줄씩 적어주는것이 괜찮다고 생각합니다!
단지 개인적인 취향일 뿐입니다.

tmdgusya commented on src/main/java/net/eno/chess/Board.java:
```diff
@@ -13,53 +13,90 @@
+            } else if (i == 3) {
+                addPiece(Color.WHITE, Piece.createPiece(Color.WHITE, Representation.QUEEN));
+                addPiece(Color.BLACK, Piece.createPiece(Color.BLACK, Representation.QUEEN));
+            } else {
```
else / else if 를 제거하지 않고는 사용할수 없을까요?

eNoLJ replied:
이 부분도 최신 코드를 봐주시면 감사하겠습니다!!

tmdgusya commented on src/main/java/net/eno/chess/Board.java:
```diff
@@ -13,53 +13,90 @@
+        this.pieces.put(Color.BLACK.toString(), new HashMap<String, List<Piece>>() {
+            {
+                put("pawn", new ArrayList<>());
+                put("piece", new ArrayList<>());
```
가독성이 좋지 않은 코드들이라고 생각됩니다.
위에서 리뷰받으신것처럼 다른 방법으로 전환해보심이 좋을듯 합니다.

eNoLJ replied:
넵 이부분도 위에 리뷰를 받아 수정했습니다!!
수정한 코드로 봐주시면 감사하겠습니다!!

tmdgusya commented on src/main/java/net/eno/chess/Board.java:
```diff
@@ -55,28 +54,28 @@ private Piece findPiece(Color color, String piece, int index) {
+        representationList.add(Representation.BISHOP);
+        representationList.add(Representation.KNIGHT);
+        representationList.add(Representation.ROOK);
+        return representationList;
```
여기서 List 로 관리하기보다는 위에서 언급되듯이 차라리 Enum 자체에서 List 로 관리했으면 어땠을까? 라는 생각이드네요.
Enum 자체에서 List 였다면 해당 코드도 필요없어지게 되겠죠?

eNoLJ replied:
이부분은 기물을 순서대로 배치하기 위해 각각의 기물을 순차적으로 리스트에 넣어 반환하도록 만들었습니다. 그리고 enum 자체가 list였다면이라는 말이 정확히 어떤식인지 좀 더 설명 가능하실까요?? 어떻게 고칠 수 있을지 감이 잘 잡히지 않습니다ㅠㅠ

tmdgusya replied:
이 부분 저도 쿠퍼때문에 알게된사실인데 
https://woowabros.github.io/tools/2017/07/10/java-enum-uses.html
이거 링크 한번 보시면 도움되실 수 있을것 같아요.

tmdgusya commented on src/main/java/net/eno/chess/Board.java:
```diff
@@ -97,4 +85,16 @@ public String showBoard() {
+    }
+
+    private Piece findPiece(Color color, String piece, int index) {
+        return this.pieces.get(color.toString()).get(piece).get(index);
```
제 개인적인 생각인데 자료구조로 인해 코드의 복잡도가 증가했다는 생각이 드는데 어떻게 생각하시나요?
차라리 색깔별로 나누고 pawn 과 일반 piece 들을 나누어서 4개의 자료구조를 이용했으면 어땠을까? 라는 생각이 듭니다.

eNoLJ replied:
이 부분도 최종적으로 수정 된걸로 봐주시면 감사하겠습니다!!

tmdgusya commented on src/main/java/net/eno/chess/Board.java:
```diff
@@ -5,57 +5,40 @@
-    private final Map<String, Map<String, List<Piece>>> pieces;
+    private final List<List<Piece>> board;
 
     public Board() {
```
미션 4 요구사항에 private 으로 바꾸는것이 있지 않았나요?

tmdgusya replied:
지금 정적 팩토리 메소드도 없는것으로 보입니다.

eNoLJ replied:
미션4 요구사항에는 Piece 클래스에 정적 팩토리 메소드를 만드는걸로 나와있더라구요 Piece class를 한번 확인 하신 후 다시 한번 조언해주시면 감사하겠습니다!!

tmdgusya commented on src/main/java/net/eno/chess/Game.java:
```diff
@@ -19,7 +21,7 @@ public static void main(String[] args) {
                 board = new Board();
                 board.initialize();
-                System.out.println(board.showBoard());
+                System.out.println(board.showBoard(Color.WHITE));
```
여기에 왜 Color.WHITE 가 들어가는걸까요..?

tmdgusya replied:
showBoard() 라는건 Color 별로 출력해주는건가요?

eNoLJ replied:
리뷰어 분께서 만약 게임판이 뒤집히게 되면, 그러니까 백팀이 위로, 흑팀이 아래로 오는 상황이라면 대처할 수 있냐는 말씀을 하셔서 그부분을 해결하기 위해 전면에 배치되는 컬러를 선택할 수 있게 만들었습니다. 흰색을 전달하면 흰색이 전면으로, 검은색을 전달하면 검은색이 전면으로 보이도록 만들었습니다.

eNoLJ replied:
아 그리고 로치 저번에도 그렇고 리뷰해주신 점 정말 감사히 생각하고 있습니다!! 많이 배우고 있어요ㅎㅎ
그치만 부탁드리고 싶은게 하나 있습니다!! 리뷰 하실 때 최신 코드로 해주시면 정말정말 더 감사할 것 같아요!!
최신 코드로 리뷰하시면 리뷰하는 로치도 그렇고 확인하는 저도 그렇고 시간절약이 가능할 것 같아요.
그래도 항상 문제가 많은 과정 중에 코드를 일일이 찝어주신거는 정말 감사하게 생각하고 있습니다!!
앞으로도 부족한 저에게 많은 조언 부탁드려요 그럼 주말 몇시간 안남았지만 푹 쉬시고 내일 또 열심히 해봐요!!

wheejuni replied:
@tmdgusya @eNoLJ 쭉 지켜봤는데요, 개인적인 생각으로는 review 기능을 사용하시기보단, 리뷰어가 제시한 리뷰 코멘트에 다시 댓글을 다는 식으로 진행하는 것이 더 좋지 않나 생각은 됩니다. 서로 의견을 나누시는 건 좋은데요... 의견 교환에 한정되었으면 좋겠고요, 지금 로치가 너무 리뷰 스타일로 코멘트를 다시는 것이 조금 우려는 됩니다.

tmdgusya replied:
@wheejuni @eNoLJ 앗 죄송합니다. 저도 커멘트나 궁금한거 물어보는 형식으로 답변하는게 좋을 것 같네요.

tmdgusya replied:
@eNoLJ 제가 달다보니 최신 수정사항을 보지 못했나보군요.. ㅠㅠ 죄송합니다.
앞으로는 궁금한거 있으면 달때 확인해서 달 수 있도록 할께요.
이노 다시한번 죄송해요

eNoLJ replied:
@tmdgusya 아니에요 로치가 죄송할 이유는 전혀 없습니다. 그렇게 생각하지 않으셨으면 좋겠습니다ㅎㅎ 한번 말씀 드렸으니 앞으로는 어떤식으로든 코멘트를 달아주셔도 감사히 받아들이겠습니다. 앞으로도 잘 부탁드려요!! 동기들 중에 제 글에 코멘트 달아주신분은 로치 밖에 없어요ㅋㅋㅋㅋ

wheejuni commented on src/main/java/net/eno/pieces/PieceType.java:
```diff
@@ -0,0 +1,24 @@
+package net.eno.pieces;
+
+public enum PieceType {
```
전반적으로 잘 구현하신 것 같네요. 👍

wheejuni commented on src/main/java/net/eno/pieces/Color.java:
```diff
@@ -0,0 +1,5 @@
+package net.eno.pieces;
+
+public enum Color {
+    WHITE, BLACK
```
한 줄에 하나씩 선언해주시는 것이 가독성 측면과 코드스타일 일관성 측면에서 조금 낫습니다.

```suggestion
    WHITE,
    BLACK

eNoLJ replied:
그렇게 하도록 하겠습니다!! 감사합니다.

wheejuni commented on src/main/java/net/eno/chess/Board.java:
```diff
@@ -1,68 +1,87 @@
 public class Board {
 
-    private final Map<String, List<Pawn>> pawns;
+    private final List<List<Piece>> board;
```
나쁘지 않은 선택이라고 생각합니다. 👍 
다만 5단계에선 이렇게 구현하시면 request change를 할 예정입니다. 
**체스판의 한 줄**에 해당하는 추상화를(클래스 설계를) 어떻게 해야할지 미리 고민해보시는 것도 좋겠네요.

eNoLJ replied:
@wheejuni 넵 저도 그 부분 인지하고 있습니다. 인지하고 있음에도 4단계에서 보충을 하지 않은거에 대해선 머라 할말이 없네요...ㅠㅠ 이번 미션4에서 많이 부족한 코드를 하나하나 조언해주셔서 감사합니다. 혹시나 미션5에서도 기회가 된다면 브라이언이 리뷰를 해주길 기대하겠습니다. 만약 또 하게 되신다면 힘드시겠지만 그때도 냉정하게 부탁드리겠습니다. 부족한 부분이 많아 사소한거 하나하나 찝어주시는거에 대해 많이 배우고 있습니다. 감사합니다!!!!

wheejuni reviewed (approved):
수고 많으셨고요, 💯  더 나아질 여지가 많이 남아있습니다만...
4단계에서 해야 할 충분한 경험을 하신 듯 하고요, 5단계로 넘어가서 더 많은 경험과 학습이 일어났으면 좋겠습니다. merge 할게요.
